GH-34971: [Format] Enhance C-Data API to support non-cpu cases (#34972)
apache/arrow: docs/source/format/CDeviceDataInterface.rst

.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. highlight:: c

.. _c-device-data-interface:

=================================
The Arrow C Device data interface
=================================

.. note:: The Arrow C Device Data Interface should be considered experimental

Rationale
=========

The current :ref:`C Data Interface <c-data-interface>`, and most
implementations of it, make the assumption that all data buffers provided
are CPU buffers. Since Apache Arrow is designed to be a universal in-memory
format for representing tabular ("columnar") data, there will be the desire
to leverage this data on Non-CPU hardware such as GPUs. One example of such

> a case is the `RAPIDS cuDF library` which uses the Arrow memory format with

Author: It does link to rapids, i'll update it to use the `:ref:` syntax but the link is at the bottom of the file.

> a case is the `RAPIDS cuDF library` which uses the Arrow memory format with

Author: sorry, not `:ref:`, but adding the `_` to indicate it. I'm still more used to md than rst lol

CUDA for NVIDIA GPUs. Since copying data from host to device and back is
expensive, the ideal would be to be able to leave the data on the device
for as long as possible, even when passing it between runtimes and
libraries.

The Arrow C Device data interface builds on the existing C data interface
by adding a very small, stable set of C definitions to it. For non-C/C++
languages and runtimes, translating the C definitions to corresponding
C FFI declarations should be just as simple as with the current C data

> interface.

Author: I do that, just later in the prose rather than right here. But I can add that extra bit here if you think it's necessary.

> interface.

Author: added a sentence here to explain. Let me know what you think

    be copied to consumer-managed storage if it is intended to survive longer.

.. c:member:: void (*ArrowDeviceArrayStream.release)(struct ArrowDeviceArrayStream*)

    *Mandatory.* A pointer to a producer-provided release callback.

.. c:member:: void* ArrowDeviceArrayStream.private_data

    *Optional.* An opaque pointer to producer-provided private data.

    Consumers MUST NOT process this member. Lifetime of this member is
    handled by the producer, and especially by the release callback.

Result lifetimes
----------------

The data returned by the ``get_schema`` and ``get_next`` callbacks must be
released independantly. Their lifetimes are not tied to that of 
``ArrowDeviceArrayStream``.

Stream lifetime
---------------

Lifetime of the C stream is managed using a release callback with similar 
usage as in :ref:`C data interface <c-data-interface-released>`.

Thread safety
-------------

The stream source is not assumed to be thread-safe. Consumers wanting to
call ``get_next`` from several threads should ensure those calls are 
serialized.

Updating this specification
===========================

Once this specification is supported in an official Arrow release, the C ABI

> is frozen. This means that the ``ArrowDeviceArray`` structure definition 

Author: Added a note here to clarify the language. Let me know your thoughts, thanks!

.. c:member:: int64_t ArrowDeviceArray.device_id

    The device id to identify a specific device if multiple devices of this
    type are on the system. The semantics of the id will be hardware dependent,
    but we use an ``int64_t`` to future-proof the id as devices change over time.

.. c:member:: ArrowDeviceType ArrowDeviceArray.device_type

    The type of the device which can access the buffers in the array.

.. c:member:: void* ArrowDeviceArray.sync_event

    Optional. An event-like object to synchronize on if needed.

    Many devices, like GPUs, are primarily asynchronous with respect to
    CPU processing. As such in order to safely access memory, it is often
    necessary to have an object to synchronize processing with. Since
    different devices will use different types to specify this, we use a
    void* which can be coerced into a pointer to whatever the device
    appropriate type is.

    If synchronization is not needed, this can be null. If this is non-null
    then it MUST be used to call the appropriate sync method for the device
    (e.g. ``cudaStreamWaitEvent`` or ``hipStreamWaitEvent``) before attempting
    to access the memory in the buffers.

>     Expected types to coerce this ``void*`` to depending on the reported

Reviewer: Perhaps detail this in a separate paragraph about event types?
For example:
```restructuredtext
Synchronization event types
---------------------------

The table below lists the event types supported for each the device type.
If no event type is supported ("N/A"), then the ``sync_event`` member
should always be null.

+--------------------------+----------------------+--------+
| Device type              | Actual event type    | Notes  |
+==========================+======================+========+
| ARROW_DEVICE_CPU         | N/A                  |        |
+--------------------------+----------------------+--------+
| ARROW_DEVICE_CUDA        | ``cudaEvent_t*``     |        |
+--------------------------+----------------------+--------+
| ...                      | ...                  |        |
+--------------------------+----------------------+--------+
```

Author: Good idea, i've updated with the table and it's own section.

Reviewer: Ok, but do we still want the bullet list below?

    device type:

    * CUDA: ``cudaEvent_t*``
    * ROCm: ``hipEvent_t*``
    * OpenCL: ``cl_event*``
    * Vulkan: ``VkEvent*``
    * Metal: ``MTLEvent*``
    * OneAPI: ``sycl::event*``

    If an event is provided, then the producer MUST ensure that the event
    is triggered/recorded at the end of the processing stream once the data

>     is considered available for use.

Reviewer: This is not clear to me. What is the "processing stream"? Can we make this simpler? e.g.:
> If an event is provided, then the producer MUST ensure that the exported data is available on the device before the even becomes ready. The consumer SHOULD wait on the event before trying to access the exported data.

Author: I've simplified and updated this, let me know if the version is good or if you think it should be further simplified.

.. c:member:: int64_t ArrowDeviceArray.reserved[3]

    As non-CPU development expands, there may be a need to expand this
    structure. In order to do so without potentially breaking ABI changes,
    we reserve 24 bytes at the end of the object. This also has the added
    benefit of bringing the total size of this structure to exactly 128
    bytes (a power of 2) on 64-bit systems. These bytes should be zero'd
    out after allocation in order to ensure safe evolution of the ABI in

>     the future.

Reviewer: I'm still skeptical about this, but if we want to keep this, replace "should" with "MUST"?

Reviewer: Also, to be clear, those should be zeroed by the producer (which is responsible for initializing the whole struct).

.. note::

>     Rather than store the shape / types of the data alongside the

Reviewer: Not sure this note is useful?

Author: One of the reasons for the creation of things such as the `Cuda Array Interface` that numba uses was in order to pass the shape and type of the data buffer along with the pointers. Since that kind of interface is what we're targeting here and we aren't putting that information into the `ArrowDeviceArray` struct, I felt it was good to explicitly state that producers should use the `ArrowSchema` struct to pass this information. But I'm open to removing this note if people feel it is unnecessary.

Reviewer: Then you should clearly state the context of this note, i.e. a comparison with the Cuda Array Interface. Otherwise it's confusing.

Perhaps you also want to move this at the end into a "Interoperability with other interchange formats" paragraph?

Author: added the new section

``ArrowSchema`` can be used for schema-level metadata of the record batch.

Mutability
----------

Both the producer and the consumer SHOULD consider the exported data (that
is, the data reachable on the device through the ``buffers`` member of
the embedded ``ArrowArray``) to be immutable, as either party could otherwise
see inconsistent data while the other is mutating it.

Likewise, if the ``sync_event`` member is non-NULL, the consumer should not
attempt to access or read the data until they have synchronized on that event.

C producer example
====================

Exporting a simple ``int32`` device array
-----------------------------------------

Export a non-nullable ``int32`` type with empty metadata. An example of this
can be seen in the :ref:`C data interface docs directly <c-data-interface-export-int32-schema>`.

To export the data itself, we transfer ownership to the consumer through
the release callback. This example will use CUDA, but the equivalent calls
could be used for any device:

.. code-block:: c

    static void release_int32_device_array(struct ArrowArray* array) {
        assert(array->n_buffers == 2);
        // destroy the event

>         cudaEvent_t* ev_ptr = reinterpret_cast<cudaEvent_t*>(array->private_data);

Reviewer: This wouldn't work since `array` is a `struct ArrowArray*` here? Need to cast to `struct ArrowDeviceArray*` first.

(note the cast might technically be UB...)

Reviewer: Also, this is sample C code, not C++, so should only use C-style casts (not `reinterpret_cast`).

Author: In this example, you can see that we store the event pointer in the `private_data` member of the `ArrowArray` member inside of `ArrowDeviceArray` and release should be called using `device_array->array.release(&device_array->array)`. That ensures that there's no UB.

Reviewer: Oh, I misread. Thanks for the clarification!

        cudaError_t status = cudaEventDestroy(*ev_ptr);
        assert(status == cudaSuccess);
        free(ev_ptr);

        // free the buffers and the buffers array
        status = cudaFree(array->buffers[1]);
        assert(status == cudaSuccess);
        free(array->buffers);

        // mark released
        array->release = NULL;
    }

>     __host__ void export_int32_device_array(void* cudaAllocdPtr,

Reviewer: If `__host__` required? If so, it should also be added to the release function above.

Reviewer 2: Given this code block is marked as `c` and there's no kernel code here we should definitely remove this as it's not needed (and likely wouldn't work unless using `nvcc` and would still likely warn).

        cudaError_t status;
        status = cudaGetDevice(&device);
        assert(status == cudaSuccess);

        cudaEvent_t* ev_ptr = (cudaEvent_t*)malloc(sizeof(cudaEvent_t));
        assert(ev_ptr != NULL);
        status = cudaEventCreate(ev_ptr);
        assert(status == cudaSuccess);

        // record event on the stream, assuming that the passed in
        // stream is where the work to produce the data will be processing.
        status = cudaEventRecord(*ev_ptr, stream);
        assert(status == cudaSuccess);

        // initialize fields
        *array = (struct ArrowDeviceArray) {
            .array = (struct ArrowArray) {
                .length = length,
                .null_count = 0,
                .offset = 0,
                .n_buffers = 2,
                .n_children = 0,
                .children = NULL,
                .dictionary = NULL,
                // bookeeping
                .release = &release_int32_device_array,
                .private_data = reinterpret_cast<void*>(ev_ptr),
            },
            .device_id = static_cast<int64_t>(device),
            .device_type = ARROW_DEVICE_CUDA,
            // pass the event pointer to the consumer

>             .sync_event = reinterpret_cast<void*>(ev_ptr),

Reviewer: Use C-style cast here.

        };

        // allocate list of buffers
        array->array.buffers = (const void**)malloc(sizeof(void*) * array->array.n_buffers);
        assert(array->array.buffers != NULL);
        array->array.buffers[0] = NULL;
        array->array.buffers[1] = cudaAllocdPtr;
    }

> ================

Reviewer: Formatting nit: should match over/underlining length with title length, otherwise I think reST won't like you :-)

released independantly. Their lifetimes are not tied to that of
``ArrowDeviceArrayStream``.

Stream lifetime
---------------

Lifetime of the C stream is managed using a release callback with similar
usage as in :ref:`C data interface <c-data-interface-released>`.

Thread safety
-------------

The stream source is not assumed to be thread-safe. Consumers wanting to
call ``get_next`` from several threads should ensure those calls are
serialized.

Updating this specification
===========================

.. note::
    Since this specification is still considered experimental, there is the
    (still very low) possibility it might change slightly. Once it is
    supported in an official Arrow release and the "experimental" tag is
    removed from it, this section will apply and the ABI will be frozen.
    
    The reason for the "experimental" tag is because we don't know what we
    don't know. While it was attempted to ensure this is generic enough to
    work with a multitude of different frameworks, it's also possible that
    something was missed. Once there is some usage of this and we are
    confident there isn't any necessary modifications, the "experimental"

>     tag will be removed and the ABI frozen.

Author: Is the new version better?

>     tag will be removed and the ABI frozen.

Reviewer: Seems good to me, thanks!

    #define ARROW_DEVICE_WEBGPU 15
    // Qualcomm Hexagon DSP
    #define ARROW_DEVICE_HEXAGON 16

    struct ArrowDeviceArray {
      struct ArrowArray array;
      int64_t device_id;
      ArrowDeviceType device_type;
      void* sync_event;

      // reserved bytes for future expansion
      int64_t reserved[3];
    };

    #endif  // ARROW_C_DEVICE_DATA_INTERFACE

.. note::
   The canonical guard ``ARROW_C_DEVICE_DATA_INTERFACE`` is meant to avoid
   duplicate definitions if two projects copy the definitions in their own
   headers, and a third-party project includes from these two projects. It
   is therefore important that this guard is kept exactly as-is when these
   definitions are copied.

ArrowDeviceType
---------------

The ``ArrowDeviceType`` typedef is used to indicate what type of device the
provided memory buffers were allocated on. This, in conjunction with the
``device_id``, should be sufficient to reference the correct data buffers.

We then use macros to define values for different device types. The provided
macro values are compatible with the widely used `dlpack`_ ``DLDeviceType``
definition values, using the same value for each as the equivalent

> ``kDL<type>`` enum from dlpack.h. The list will be kept in sync with those

Reviewer: ```suggestion
``kDL<type>`` enum from ``dlpack.h``. The list will be kept in sync with those
```

added to dlpack before we add a corresponding macro here.

To ensure predictability with the ABI, we use macros instead of an ``enum``
so the storage type is not compiler dependent.

.. c:macro:: ARROW_DEVICE_CPU

    CPU Device, equivalent to just using ``ArrowArray`` directly instead of
    using ``ArrowDeviceArray``.

.. c:macro:: ARROW_DEVICE_CUDA

    A `CUDA`_ GPU Device. This could represent data allocated either with the
    runtime library (``cudaMalloc``) or the device driver (``cuMemAlloc``).

.. c:macro:: ARROW_DEVICE_CUDA_HOST

    CPU memory that was pinned and page-locked by CUDA by using
    ``cudaMallocHost`` or ``cuMemAllocHost``.

.. c:macro:: ARROW_DEVICE_OPENCL

    Data allocated on the device by using the `OpenCL (Open Computing Language)`_
    framework.

.. c:macro:: ARROW_DEVICE_VULKAN

    Data allocated by the `Vulkan`_ framework and libraries.

.. c:macro:: ARROW_DEVICE_METAL

    Data on Apple GPU devices using the `Metal`_ framework and libraries.

.. c:macro:: ARROW_DEVICE_VPI

    Indicates usage of a Verilog simulator buffer.

.. c:macro:: ARROW_DEVICE_ROCM

    An AMD device using the `ROCm`_ stack.

.. c:macro:: ARROW_DEVICE_ROCM_HOST

>     CPU memory pinned and page-locked allocated ``hipMallocHost``.

Reviewer: ```suggestion
    CPU memory that was pinned and page-locked by ROCm by using ``hipMallocHost``.
```

.. c:macro:: ARROW_DEVICE_EXT_DEV

    This value is an escape-hatch for devices to extend which aren't
    currently represented otherwise. Producers would need to provide
    additional information/context specific to the device if using
    this device type. This is used to quickly test extension devices
    and semantics can differ based on the implementation.

.. c:macro:: ARROW_DEVICE_CUDA_MANAGED

    CUDA managed/unified memory which is allocated by ``cudaMallocManaged``.

.. c:macro:: ARROW_DEVICE_ONEAPI

    Unified shared memory allocated on an Intel `oneAPI`_ non-partitioned
    device. A call to the ``oneAPI`` runtime is required to determine
    the specific device type, the USM allocation type and the sycl context
    that it is bound to.

.. c:macro:: ARROW_DEVICE_WEBGPU

    GPU support for next-gen WebGPU standards

.. c:macro:: ARROW_DEVICE_HEXAGON

    Data allocated on a Qualcomm Hexagon DSP device.

The ArrowDeviceArray structure
------------------------------

The ``ArrowDeviceArray`` structure embeds the C data ``ArrowArray`` structure
and adds additional information necessary for consumers to use the data. It
has the following fields:

> .. c:member:: struct ArrowArray ArrowDeviceArray.array

Reviewer: For each of these, define if it's mandatory or optional.

Author: done

    The allocated array data. The values in the ``void**`` buffers (along
    with the buffers of any children) are what is allocated on the device.
    The buffer values should be device pointers. The rest of the structure
    should be accessible to the CPU.

    The ``private_data`` and ``release`` callback of this structure should
    contain any necessary information and structures related to freeing
    the array according to the device it is allocated on, rather than
    having a separate release callback and ``private_data`` pointer here.

.. c:member:: int64_t ArrowDeviceArray.device_id

    The device id to identify a specific device if multiple devices of this
    type are on the system. The semantics of the id will be hardware dependent,
    but we use an ``int64_t`` to future-proof the id as devices change over time.

.. c:member:: ArrowDeviceType ArrowDeviceArray.device_type

    The type of the device which can access the buffers in the array.

.. c:member:: void* ArrowDeviceArray.sync_event

    Optional. An event-like object to synchronize on if needed.

    Many devices, like GPUs, are primarily asynchronous with respect to

>     CPU processing. As such in order to safely access memory, it is often

Reviewer: ```suggestion
    CPU processing. As such, in order to safely access device memory, it is often
```

    necessary to have an object to synchronize processing with. Since
    different devices will use different types to specify this, we use a

>     void* which can be coerced into a pointer to whatever the device

Reviewer: ```suggestion
    ``void*`` which can be coerced into a pointer to whatever the device
```

    then it MUST be used to call the appropriate sync method for the device
    (e.g. ``cudaStreamWaitEvent`` or ``hipStreamWaitEvent``) before attempting
    to access the memory in the buffers.

    Expected types to coerce this ``void*`` to depending on the reported
    device type:

    * CUDA: ``cudaEvent_t*``
    * ROCm: ``hipEvent_t*``
    * OpenCL: ``cl_event*``
    * Vulkan: ``VkEvent*``
    * Metal: ``MTLEvent*``
    * OneAPI: ``sycl::event*``

    If an event is provided, then the producer MUST ensure that the exported
    data is available on the device before the event is triggered. The 
    consumer SHOULD wait on the event before trying to access the exported
    data.

.. c:member:: int64_t ArrowDeviceArray.reserved[3]

    As non-CPU development expands, there may be a need to expand this
    structure. In order to do so without potentially breaking ABI changes,
    we reserve 24 bytes at the end of the object. This also has the added
    benefit of bringing the total size of this structure to exactly 128
    bytes (a power of 2) on 64-bit systems. These bytes MUST be zero'd
    out after initialization by the producer in order to ensure safe 

>     evolution of the ABI in the future.

Reviewer: I would not mention the dubious "added benefit" of making the size a power of two.

+---------------------------+--------------------+---------+
| ARROW_DEVICE_ROCM         | ``hipEvent_t*``    |         |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_ROCM_HOST    | ``hipEvent_t*``    |         |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_EXT_DEV      |                    | (2)     |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_CUDA_MANAGED | ``cudaEvent_t*``   |         |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_ONEAPI       | ``sycl::event*``   |         |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_WEBGPU       | N/A                | (1)     |
+---------------------------+--------------------+---------+
| ARROW_DEVICE_HEXAGON      | N/A                | (1)     |
+---------------------------+--------------------+---------+

Notes:

* \(1) Currently unknown if framework has an event type to support.
* \(2) Extension Device has producer defined semantics and thus if
       synchronization is needed for an extension device, the producer
       should document the type.

Semantics

> =========

Reviewer: Perhaps we should make clear somewhere that, amongst all the information exported by the C Device Data Interface, only the data buffers reside in device memory, all the rest is in CPU memory.

Author: added at the top of the `Memory Management` section.

        cudaError_t status = cudaEventDestroy(*ev_ptr);
        assert(status == cudaSuccess);
        free(ev_ptr);

        // free the buffers and the buffers array
        status = cudaFree(array->buffers[1]);
        assert(status == cudaSuccess);
        free(array->buffers);

        // mark released
        array->release = NULL;
    }

    void export_int32_device_array(void* cudaAllocdPtr,
                                   cudaStream_t stream,
                                   int64_t length,
                                   struct ArrowDeviceArray* array) {
        // get device id
        int device;
        cudaError_t status;
        status = cudaGetDevice(&device);
        assert(status == cudaSuccess);

        cudaEvent_t* ev_ptr = (cudaEvent_t*)malloc(sizeof(cudaEvent_t));
        assert(ev_ptr != NULL);
        status = cudaEventCreate(ev_ptr);
        assert(status == cudaSuccess);

        // record event on the stream, assuming that the passed in
        // stream is where the work to produce the data will be processing.
        status = cudaEventRecord(*ev_ptr, stream);
        assert(status == cudaSuccess);

        // initialize fields

>         *array = (struct ArrowDeviceArray) {

Reviewer: Should `memset` `*array` before, or explicitly zero out the reserved area.

Author: added

            .array = (struct ArrowArray) {
                .length = length,
                .null_count = 0,
                .offset = 0,
                .n_buffers = 2,
                .n_children = 0,
                .children = NULL,
                .dictionary = NULL,

>                 // bookeeping

Reviewer: ```suggestion
                // bookkeeping
```

                .release = &release_int32_device_array,
                // store the event pointer as private data in the array
                // so that we can access it in the release callback.
                .private_data = (void*)(ev_ptr),
            },
            .device_id = (int64_t)(device),
            .device_type = ARROW_DEVICE_CUDA,
            // pass the event pointer to the consumer
            .sync_event = (void*)(ev_ptr),
        };

        // allocate list of buffers
        array->array.buffers = (const void**)malloc(sizeof(void*) * array->array.n_buffers);
        assert(array->array.buffers != NULL);
        array->array.buffers[0] = NULL;
        array->array.buffers[1] = cudaAllocdPtr;
    }

    // calling the release callback should be done using the array member
    // of the device array.
    static void release_device_array_helper(struct ArrowDeviceArray* arr) {
        arr->array.release(&arr->array);
    }

=======================
Device Stream Interface
=======================

Like the :ref:`C stream interface <c-stream-interface>`, the C Device data
interface also specifies a higher-level structure for easing communication

> of streaming data within a single process. Defining an ``ArrowDeviceArrayStream``
> structure.

Reviewer: Is there a leftover sentence here?
```suggestion
of streaming data within a single process.
```

        // callbacks
        int (*get_schema)(struct ArrowDeviceArrayStream*, struct ArrowSchema*);
        int (*get_next)(struct ArrowDeviceArrayStream*, struct ArrowDeviceArray*);
        const char* (*get_last_error)(struct ArrowDeviceArrayStream*);

        // release callback
        void (*release)(struct ArrowDeviceArrayStream*);

        // opaque producer-specific data
        void* private_data;
    };

    #endif  // ARROW_C_DEVICE_STREAM_INTERFACE

.. note::
    The canonical guard ``ARROW_C_DEVICE_STREAM_INTERFACE`` is meant to avoid
    duplicate definitions if two projects copy the C device stream interface
    definitions into their own headers, and a third-party project includes
    from these two projects. It is therefore important that this guard is
    kept exactly as-is when these definitions are copied.

The ArrowDeviceArrayStream structure
------------------------------------

The ``ArrowDeviceArrayStream`` provides a device type that can access the
resulting data along with the required callbacks to interact with a
streaming source of Arrow arrays. It has the following fields:

.. c:member:: ArrowDeviceType device_type

>     The device type that this stream produces data on. All

Reviewer: Mandatory or optional?
